### ISSUE TITLE
Window function fusion for value window functions

### ISSUE THREAD

**Issue by ggevay** [MaterializeInc/materialize#29425]
**This issue was created after the original window function fusion issue was turned into an epic of various window function optimizations. For discussion, including links to various users needing this, see [that issue](https://github.com/MaterializeInc/materialize/issues/20741).** 

If a query has several window function calls with the same PARTITION BY ([a big user has a query with 8!](https://materializeinc.slack.com/archives/C04SN4BEQ0N/p1689965070747669)), then we could bundle them into one call to amortize various overheads (e.g., call `reduce` only once). (Different `ORDER BY`s are ok.) This is similar to how we bundle several aggregations in the same query, but the difference is that we need to do this for window functions at the HIR level. We could create a new window function called `bundled_win_funcs`, which would have a `Vec` of window function calls.

Note that fixing this would improve both CPU and memory usage. It would help memory because currently we arrange independent columns (columns that are not involved in the window function calls) several times.

Without fusion:
```
CREATE TABLE t(y int);

explain select y, lag(y+100) over (), lag(y+200) over () from t;
                                                                                       Optimized Plan                                                                                        
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Explained Query:                                                                                                                                                                           +
   Project (#3, #5, #4)                                                                                                                                                                     +
     Map (record_get[1](#1), record_get[0](#2), record_get[2](#2), record_get[0](#1))                                                                                                       +
       FlatMap unnest_list(#0)                                                                                                                                                              +
         Reduce aggregates=[lag[order_by=[]](row(row(row(record_get[0](record_get[1](#0)), record_get[0](#0), record_get[0](#0)), row((record_get[0](record_get[1](#0)) + 100), 1, null))))]+
           Project (#1)                                                                                                                                                                     +
             FlatMap unnest_list(#0)                                                                                                                                                        +
               Reduce aggregates=[lag[order_by=[]](row(row(row(#0), row((#0 + 200), 1, null))))]                                                                                            +
                 ReadStorage materialize.public.t                                       
```

Note that https://github.com/MaterializeInc/materialize/pull/29276 solved the value window function case, which was blocking https://github.com/MaterializeInc/accounts/issues/3. I've created a separate issue for window aggregations: https://github.com/MaterializeInc/materialize/issues/29478

**Comment by ggevay** [comment]
Split off the window aggregation case. Closing, because it's done for value window functions in https://github.com/MaterializeInc/materialize/pull/29276
